lib/spack/spack/spec.py

```diff
@@ -1958,16 +1957,29 @@ def concretize(self, tests=False):
                     for pcond, patch_list in dependency.patches.items():
                         if dspec.spec.satisfies(pcond):
                             for patch in patch_list:
-                                patches.append(patch.sha256)
+                                patches.append(patch)
             if patches:
-                mvar = dspec.spec.variants.setdefault(
-                    'patches', MultiValuedVariant('patches', ())
-                )
-                mvar.value = mvar.value + tuple(patches)
-                # FIXME: Monkey patches mvar to store patches order
-                p = getattr(mvar, '_patches_in_order_of_appearance', [])
-                mvar._patches_in_order_of_appearance = list(
-                    dedupe(p + patches))
+                all_patches = spec_to_patches.setdefault(id(dspec.spec), [])
+                all_patches.extend(patches)
+
+        for spec in self.traverse():
+            if id(spec) not in spec_to_patches:
+                continue
+
+            patches = list(dedupe(spec_to_patches[id(spec)]))
+            mvar = spec.variants.setdefault(
+                'patches', MultiValuedVariant('patches', ())
+            )
+            mvar.value = tuple(p.sha256 for p in patches)
```

Comment:
Wouldn't it be cleaner to just store the patches in order in the first place? Or is there some reason I'm not seeing that this couldn't work.

`mvar.value = tuple(p.sha256 for p in sorted(patches, key=lambda p: (p.ordering_key, p.sha256)))`

That way I don't think we would need to use the monkey-patch.

Reply:
The `_patches_in_order_of_appearance` field was added by https://github.com/spack/spack/commit/5fa1191d1736659f03b3e49bf35db42b576d2266, and I think the concern of that edit still applies here.

Note that the setter for the `variant.value` applies sorting and deduplication itself in order to maintain consistency of `__hash__` (which is used for `cmp_key` rather than `Spec.dag_hash`). https://github.com/spack/spack/pull/5596#discussion_r142757461 may shed more light as well.

Reply:
Ok, I hadn't realized the variant would reorder them.

In the long run, we should change `Patch` objects to have a `cmp_key` and set that based on their order, so they end up sorted properly in `variant.value`. But I think it's fair to keep that outside the scope of this PR.